src/lib.rs

```diff
@@ -220,8 +215,50 @@ async fn keyboard_demo() {
             DecodedKey::RawKey(key) => tracing::warn!(?key, "you typed something weird"),
         }
         if newline {
-            shell::eval(&line);
-            line.clear();
+            return line;
+        }
+    }
+}
+
+async fn serial_line(port: usize, mut line: alloc::string::String) -> alloc::string::String {
+    let serial_iface = &crate::drivers::serial_input::SERIAL_INPUTS.get()[port];
+
+    loop {
+        let b = serial_iface.next_byte().await;
+        let mut newline = false;
+        match b {
+            b'\n' | b'\r' => {
+                newline = true;
+            }
+            // backspace
+            0x7f => {
+                line.pop();
+            }
+            c @ b'a'..=b'z' | c @ b'A'..=b'Z' | c @ b' ' => line.push(c as char),
+            other => tracing::warn!(?other, "you typed something weird"),
+        }
+        if newline {
+            return line;
         }
     }
 }
+
+type ReadLineFn = fn(alloc::string::String) -> core::pin::Pin<
+    alloc::boxed::Box<dyn core::future::Future<Output = alloc::string::String> + Send>
```

Comment:
boxing a future for every line makes me a bit sad since i know my allocator isn't very good...

Reply:
oh god i typed Box and still forgot that that means an alloc anyway for each line. damnit

anyway probably? i'll take another shot at making it more generator-y where you give it a borrow and it eventually yields `Ok(())` when the line is done or something.